packages/components/src/confirm-dialog/stories/index.js

```diff
@@ -121,22 +56,103 @@ export const Controlled = () => {
 		setConfirmVal( 'Cancelled' );
 		setIsOpen( false );
 	};
-
 	return (
 		<>
+			<Button variant="primary" onClick={ () => setIsOpen( true ) }>
+				Open ConfirmDialog
+			</Button>
+
 			<ConfirmDialog
 				isOpen={ isOpen }
 				onConfirm={ handleConfirm }
 				onCancel={ handleCancel }
+				cancelButtonText={ args.cancelButtonText }
+				confirmButtonText={ args.confirmButtonText }
 			>
-				{ daText() }
+				{ args.children }
 			</ConfirmDialog>
 
 			<Heading level={ 1 }>{ confirmVal }</Heading>
-
-			<Button variant="primary" onClick={ () => setIsOpen( true ) }>
-				Open ConfirmDialog
-			</Button>
 		</>
 	);
 };
+
+// Simplest usage: just declare the component with the required `onConfirm` prop. Note: the `onCancel` prop is optional here, unless you'd like to render the component in Controlled mode (see below)
+export const _default = Template.bind( {} );
+const _defaultSnippet = `() => {
+  const [ isOpen, setIsOpen ] = useState( false );
+  const [ confirmVal, setConfirmVal ] = useState('');
+
+  const handleConfirm = () => {
+    setConfirmVal( 'Confirmed!' );
+    setIsOpen( false );
+  };
+
+  const handleCancel = () => {
+    setConfirmVal( 'Cancelled' );
+    setIsOpen( false );
+  };
+
+  return (
+    <>
+    <ConfirmDialog
+      isOpen={ isOpen }
+      onConfirm={ handleConfirm }
+      onCancel={ handleCancel }
+    >
+      Would you like to privately publish the post now?
+    </ConfirmDialog>
+
+    <Heading level={ 1 }>{ confirmVal }</Heading>
+
+    <Button variant="primary" onClick={ () => setIsOpen( true ) }>
+      Open ConfirmDialog
+    </Button>
+    </>
+    );
+  };`;
+_default.args = {};
+_default.parameters = {
+	docs: {
+		source: {
+			code: _defaultSnippet,
+			language: 'jsx',
+			type: 'auto',
+			format: 'true',
+		},
+	},
+};
+
+// To customize button text, pass the `cancelButtonText` and/or `confirmButtonText` props.
+export const withCustomButtonLabels = Template.bind( {} );
+withCustomButtonLabels.args = {
+	cancelButtonText: 'No thanks',
+	confirmButtonText: 'Yes please!',
+};
+
+const uncontrolledSnippet = `<>
+  <ConfirmDialog
+    onConfirm={ handleConfirm }
+  >
+    Would you like to privately publish the post now?
+  </ConfirmDialog>
+  
+  <Heading level={ 1 } />
+</>
+`;
+
+export const Uncontrolled = Template.bind( {} );
+Uncontrolled.args = {};
+Uncontrolled.parameters = {
+	docs: {
+		description: {
+			story:
+				'To render in Uncontrolled Mode, omit passing a boolean to the `isOpen` prop. This will allow the component to close itself without the need for an explicit callback. In Uncontrolled mode, `onCancel` is optional.',
+		},
+		source: {
+			code: uncontrolledSnippet,
+			language: 'jsx',
+			type: 'auto',
+		},
+	},
+};
```

Comment:
Looking at this, I'm not actually sure that we should have an uncontrolled example:

- this uncontrolled example wouldn't really add any value to the  would be probably misleading (it uses the "controlled" version of the component to showcase the uncontrolled?)
- the snippet, in its current status, is somewhat incomplete (`handleConfirm` is not defined anywhere, the `Heading` component doesn't have text and is not relevant to the example)

I still think that we should somehow document the uncontrolled mode, though. Currently, [we do mention it in the README](https://github.com/WordPress/gutenberg/blob/aca7e6ee495cf6cbbab283ab682c1b81c9070797/packages/components/src/confirm-dialog/README.md#uncontrolled-mode), but we should consider how we can do that effectively in Storybook if we eventually want to move all of our docs.

@mirka what do you think?

Reply:
Given that the conditional props will [prevent](https://github.com/WordPress/gutenberg/pull/40164#issuecomment-1111466890) us from getting a clean props table anyway, it might be worth considering splitting it into separate components (e.g. `ConfirmDialog` and `ConfirmDialog.Uncontrolled`?), or maybe even removing the uncontrolled version.

Reply:
That's an interesting point.

For the sake of this PR, I guess we could remove the uncontrolled example completely from this PR, and tackle the conditional props and the uncontrolled component in a separate PR

Reply:
I've just looked over the four current implementations of `ConfirmDialog` in Gutenberg, and currently they're all Controlled, which lends more weight to the idea of doing away with Uncontrolled altogether.

I agree with removing the Uncontrolled story for now at the very least, and we can perhaps update the component to only have the single Controlled mode in a followup - which would also eliminate the conditional props at the same time which would be a double-win.